lib/services/local/session.go

```diff
@@ -95,3 +97,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
+	if err := req.Check(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	item, err := r.backend.Get(ctx, webTokenKey(req.Token))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	token, err := types.UnmarshalWebToken(item.Value, services.SkipValidation())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return token, nil
+}
+
+// List gets all web tokens.
+func (r *webTokens) List(ctx context.Context) (out []types.WebToken, err error) {
+	key := backend.Key(webPrefix, tokensPrefix)
+	result, err := r.backend.GetRange(ctx, key, backend.RangeEnd(key), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, item := range result.Items {
+		token, err := types.UnmarshalWebToken(item.Value, services.SkipValidation())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		out = append(out, token)
+	}
+	return out, nil
+}
+
+// Upsert updates the existing or inserts a new web token.
+func (r *webTokens) Upsert(ctx context.Context, token types.WebToken) error {
+	bytes, err := types.MarshalWebToken(token, services.WithVersion(services.V1))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	metadata := token.GetMetadata()
+	item := backend.Item{
+		Key:     webTokenKey(token.GetToken()),
+		Value:   bytes,
+		Expires: metadata.Expiry(),
+	}
+	_, err = r.backend.Put(ctx, item)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// Delete deletes the web token specified with req from the storage.
+func (r *webTokens) Delete(ctx context.Context, req types.DeleteWebTokenRequest) error {
+	if err := req.Check(); err != nil {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(r.backend.Delete(ctx, webTokenKey(req.Token)))
+}
+
+// DeleteAll removes all web tokens.
+func (r *webTokens) DeleteAll(ctx context.Context) error {
+	startKey := backend.Key(webPrefix, tokensPrefix)
+	if err := r.backend.DeleteRange(ctx, startKey, backend.RangeEnd(startKey)); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+type webTokens struct {
+	backend backend.Backend
+	log     logrus.FieldLogger
+}
+
+// DELETE in 7.x.
+// getLegacySession returns the web session for the specified user/sessionID
+// under a legacy path /web/users/<user>/sessions/<id>
+func getLegacyWebSession(ctx context.Context, backend backend.Backend, user, sessionID string) (services.WebSession, error) {
+	item, err := backend.Get(ctx, legacyWebSessionKey(user, sessionID))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	session, err := services.GetWebSessionMarshaler().UnmarshalWebSession(item.Value, services.SkipValidation())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// this is for backwards compatibility to ensure we
+	// always have these values
+	session.SetUser(user)
+	session.SetName(sessionID)
+	return session, nil
+}
+
+func webSessionKey(sessionID string) (key []byte) {
+	return backend.Key(webPrefix, sessionsPrefix, sessionID)
```

Comment:
So with this change to store all sessions under the same key now, would it still be possible, for example, to implement features like "remove all sessions for a particular user" if we wanted? It'd just require iterating over all sessions?

Reply:
Yes, the downside is that it requires to iterate over all values in order to remove session only for a specific user. As the web/app sessions (and tokens) are explicitly expired, this is not such a big deal but might need to be fixed nonetheless at some point.